services/src/main/java/org/keycloak/protocol/oidc/utils/OAuth2CodeParser.java

```diff
@@ -78,51 +79,57 @@ public static String persistCode(KeycloakSession session, AuthenticatedClientSes
      */
     public static ParseResult parseCode(KeycloakSession session, String code, RealmModel realm, EventBuilder event) {
         ParseResult result = new ParseResult(code);
-
         String[] parsed = DOT.split(code, 3);
         if (parsed.length < 3) {
             logger.warn("Invalid format of the code");
             return result.illegalCode();
         }
-
+        String codeUUID = parsed[0];
         String userSessionId = parsed[1];
         String clientUUID = parsed[2];
 
         event.detail(Details.CODE_ID, userSessionId);
         event.session(userSessionId);
 
-        // Parse UUID
-        String codeUUID;
-        try {
-            codeUUID = parsed[0];
-        } catch (IllegalArgumentException re) {
-            logger.warn("Invalid format of the UUID in the code");
-            return result.illegalCode();
-        }
+        SingleUseObjectProvider codeStore = session.getProvider(SingleUseObjectProvider.class);
+        Map<String, String> codeData = codeStore.remove(codeUUID);
```

Comment:
I wonder if this should be part of the retriable block, so that once that is complete the removal of the code is already committed to the DB.

Reply:
It was originally part of the retriable block, but I started hitting RETRY_SERIALIZABLE errors in the SingleUseObjectStore..

Reply:
Hm, still makes me a bit uncomfortable. In the other places we wrapped almost the whole REST endpoint to be retriable. Could we do the same here, that is: wrapping the contents of `TokenEndpoint.codeToToken()`?

Reply:
I can try it, but I would need to pass the session as a parameter to codeToToken(), and maybe to some other pieces of the code that are using the session injected into the token endpoint.

Reply:
Gave it a try - https://github.com/keycloak/keycloak/commit/a14737e786048aa5d619a1a1724f88b1b931b036 

Test passed with this approach as well, but I didn't run the whole testsuite to ensure it isn't introducing regressions somewhere else. If you think this is a preferred approach I can run the testsuite and see how it goes.

Reply:
@sguilhen - thanks for trying this one. I'll then approve this one. Still I'd like you to bring this up in our storage team meeting so we can discuss if there should be a new issue to investigate this.